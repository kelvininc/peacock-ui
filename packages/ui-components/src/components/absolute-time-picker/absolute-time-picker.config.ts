export const DEFAULT_HEADER_TITLE = 'Custom Interval';

export const DATE_INPUT_PLACEHOLDER = 'dd-mm-yyyy 00:00:00';
export const DATETIME_INPUT_MASK = 'DD-MM-YYYY HH:mm:ss';
export const DATE_INPUT_MASK = 'DD-MM-YYYY';
export const CALENDAR_MASK = 'YYYY-MM-DD';
export const CALENDAR_DATE_TIME_MASK = 'YYYY-MM-DD HH:mm:ss';
