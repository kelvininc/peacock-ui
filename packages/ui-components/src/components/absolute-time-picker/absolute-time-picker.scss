@import '../../assets/styles/globals';

.absolute-time-picker-container {
	display: flex;
	flex-direction: column;
	gap: $spacing-3x;
	background-color: kv-color('neutral-7');
	padding: $spacing-4x;
	border-radius: 4px;
	width: fit-content;

	.navigate-back {
		display: flex;
		gap: $spacing;
		align-items: center;
		cursor: pointer;
		width: fit-content;

		kv-icon {
			--icon-color: #{kv-color('neutral-4')};
		}

		.back-text {
			@include kv-font-label-small-regular;
			color: kv-color('neutral-4');
			user-select: none;
		}
	}

	.header {
		.title {
			@include kv-font-h4-semibold;
			color: kv-color('neutral-2');
		}
	}

	.absolute-range-input {
		display: flex;
		gap: $spacing-6x;

		kv-date-time-input {
			width: 196px;
		}
	}

	.absolute-point-input {
		width: 100%;
	}

	.calendars {
		display: flex;
		gap: $spacing-6x;
	}
}
