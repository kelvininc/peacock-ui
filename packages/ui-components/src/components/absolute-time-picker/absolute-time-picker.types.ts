import { EventEmitter } from '@stencil/core';
import { EAbsoluteTimeError, EValidationState, SelectedRange } from '../../types';

export enum EAbsoluteTimePickerMode {
	Single = 'single',
	Range = 'range'
}

export interface IAbsoluteTimePicker {
	/** (optional) Defines if the calendar is in single date or range mode */
	mode?: EAbsoluteTimePickerMode;
	/** (optional) Title disaplayed on top of the component */
	headerTitle?: string;
	/** (optional) Enables the back button displayed on top */
	displayBackButton?: boolean;
	/** (optional) Selected dates */
	selectedDates?: string[];
	/** (optional) Initial Date */
	initialDate?: string;
	/** (optional) Disabled dates */
	disabledDates?: string[];
	/** (optional) use to determine if the chart inputs have custom strings */
	relativeTimeConfig?: IRelativeTimeInput;
	/** (optional) calendar minimum date to be navigated format: DD-MM-YYYY HH:mm:ss */
	calendarInputMinDate?: string;
	/** (optional) calendar maximum date to be navigated format: DD-MM-YYYY HH:mm:ss */
	calendarInputMaxDate?: string;
	/** (optional) It is used to determine if the time picker as date time input error */
	error?: EAbsoluteTimeError;
}

export interface IAbsoluteTimePickerEvents {
	/** Selected dates change */
	selectedDatesChange: EventEmitter<IAbsoluteSelectedRangeDates>;
	/** Emitted when the back button is clicked */
	backButtonClicked: EventEmitter<MouseEvent>;
	/** Emitted when the input is clicked and it were displaying custom text */
	relativeTimeConfigReset: EventEmitter<MouseEvent>;
	/** Emitted when there is a change in the relative config */
	relativeTimeConfigChange: EventEmitter<IAbsoluteSelectedRangeDates>;
}

export interface IAbsoluteSelectedRangeDates {
	range: SelectedRange;
}

export interface IRelativeTimeInput {
	mode: ERelativeTimeInputMode;
	from: string;
	to: string;
}

export enum ERelativeTimeInputMode {
	Text = 'text',
	Date = 'date'
}

export enum EInputSource {
	From = 'from',
	To = 'to',
	Single = 'single'
}

export interface DateInputState {
	state: EValidationState;
	helpText: string;
}
