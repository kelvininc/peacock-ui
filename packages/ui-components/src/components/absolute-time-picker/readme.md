# kv-absolute-time-picker



<!-- Auto Generated Below -->


## Usage

### Angular

```html
<!-- Default -->
<kv-absolute-time-picker></kv-absolute-time-picker>

<!-- With selected ranges date -->
<kv-absolute-time-picker [selectedDates]="['2022-08-04', '2022-09-01']"></kv-absolute-time-picker>

<!-- With initial date -->
<kv-absolute-time-picker initialDate="2022-08-04"></kv-absolute-time-picker>

<!-- With disabled dates-->
<kv-absolute-time-picker [disabledDates]="['2021-12-01', '2021-12-04', '2021-12-05']"></kv-absolute-time-picker>

<!-- With min and max date -->
<kv-absolute-time-picker minDate="2021-12-04" maxDate="2023-02-12"></kv-absolute-time-picker>
```


### React

```tsx
import React from 'react';

import { KvAbsoluteTimePicker } from '@kelvininc/react-ui-components';

export const KvAbsoluteTimePicker: React.FC = () => (
	<>
		{/*-- Default --*/}
		<KvAbsoluteTimePicker />

		{/*-- With selected date --*/}
		<KvAbsoluteTimePicker selectedDates={['2022-08-04', '2022-09-01']} />

		{/*-- With initial date --*/}
		<KvAbsoluteTimePicker initialDate="2021-12-03" />

		{/*-- With disabled dates --*/}
		<KvAbsoluteTimePicker disabledDates={['2021-12-01', '2021-12-04', '2021-12-05']} />

		{/*-- With min and max --*/}
		<KvAbsoluteTimePicker minDate="2021-12-04" maxDate="2023-02-12" />
	</>
);
```



## Properties

| Property               | Attribute                 | Description                                                                  | Type                                                              | Default                         |
| ---------------------- | ------------------------- | ---------------------------------------------------------------------------- | ----------------------------------------------------------------- | ------------------------------- |
| `calendarInputMaxDate` | `calendar-input-max-date` | (optional) calendar maximum date to be navigated format: DD-MM-YYYY HH:mm:ss | `string`                                                          | `CALENDAR_INPUT_MAX_DATE`       |
| `calendarInputMinDate` | `calendar-input-min-date` | (optional) calendar minimum date to be navigated format: DD-MM-YYYY HH:mm:ss | `string`                                                          | `CALENDAR_INPUT_MIN_DATE`       |
| `disabledDates`        | --                        | (optional) Disabled dates                                                    | `string[]`                                                        | `[]`                            |
| `displayBackButton`    | `display-back-button`     | (optional) Enables the back button displayed on top                          | `boolean`                                                         | `false`                         |
| `headerTitle`          | `header-title`            | (optional) Title disaplayed on top of the component                          | `string`                                                          | `DEFAULT_HEADER_TITLE`          |
| `initialDate`          | `initial-date`            | (optional) Initial Date                                                      | `string`                                                          | `undefined`                     |
| `mode`                 | `mode`                    | (optional) Defines if the calendar is in single date or range mode           | `EAbsoluteTimePickerMode.Range \| EAbsoluteTimePickerMode.Single` | `EAbsoluteTimePickerMode.Range` |
| `relativeTimeConfig`   | --                        | (optional) use to determine if the chart inputs have custom strings          | `IRelativeTimeInput`                                              | `undefined`                     |
| `selectedDates`        | --                        | (optional) Selected dates                                                    | `string[]`                                                        | `[]`                            |


## Events

| Event                      | Description                                                          | Type                                       |
| -------------------------- | -------------------------------------------------------------------- | ------------------------------------------ |
| `backButtonClicked`        | Emitted when the back button is clicked                              | `CustomEvent<MouseEvent>`                  |
| `relativeTimeConfigChange` | Emitted when there is a change in the relative config                | `CustomEvent<IAbsoluteSelectedRangeDates>` |
| `relativeTimeConfigReset`  | Emitted when the input is clicked and it were displaying custom text | `CustomEvent<MouseEvent>`                  |
| `selectedDatesChange`      | Selected dates change                                                | `CustomEvent<IAbsoluteSelectedRangeDates>` |


## Dependencies

### Used by

 - [kv-absolute-time-picker-dropdown](../absolute-time-picker-dropdown)
 - [kv-time-picker](../time-picker)

### Depends on

- [kv-icon](../icon)
- [kv-date-time-input](../date-time-input)
- [kv-time-picker-calendar](../time-picker-calendar)

### Graph
```mermaid
graph TD;
  kv-absolute-time-picker --> kv-icon
  kv-absolute-time-picker --> kv-date-time-input
  kv-absolute-time-picker --> kv-time-picker-calendar
  kv-date-time-input --> kv-form-label
  kv-time-picker-calendar --> kv-icon
  kv-time-picker-calendar --> kv-time-picker-calendar-day
  kv-absolute-time-picker-dropdown --> kv-absolute-time-picker
  kv-time-picker --> kv-absolute-time-picker
  style kv-absolute-time-picker fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


