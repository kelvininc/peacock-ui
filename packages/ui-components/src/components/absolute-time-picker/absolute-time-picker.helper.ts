import dayjs from 'dayjs';
import { CALENDAR_DATE_TIME_MASK, CALENDAR_MASK, DATETIME_INPUT_MASK } from './absolute-time-picker.config';
import { DateInputState, EAbsoluteTimeError, EValidationState, IAbsoluteTimeLimits, SelectedRange } from '../../types';

export const buildSelectedDatesEventPayload = (dateA?: dayjs.Dayjs, dateB?: dayjs.Dayjs): SelectedRange => {
	if (!dateA && !dateB) {
		return [];
	}

	if (!dateB) {
		return [dateA.format(CALENDAR_DATE_TIME_MASK)];
	}

	return [dateA.format(CALENDAR_DATE_TIME_MASK), dateB.format(CALENDAR_DATE_TIME_MASK)];
};

export const isEndDateAtStartOfDay = (date: dayjs.Dayjs): boolean => {
	return date.isValid() && date.hour() === 0 && date.minute() === 0 && date.second() === 0;
};

export const getFirstCalendarInitialDate = (displayedMonth: dayjs.Dayjs): string => {
	const initialDate = displayedMonth;
	return initialDate.isValid() ? initialDate.format(CALENDAR_MASK) : '';
};

export const getSecondCalendarInitialDate = (displayedMonth: dayjs.Dayjs): string => {
	const initialDate = displayedMonth;
	return initialDate.isValid() ? initialDate.add(1, 'month').format(CALENDAR_MASK) : '';
};

export const getMinimumDateFromDayClick = (clickedDate: dayjs.Dayjs, minimumDate: string): dayjs.Dayjs => {
	const parsedMinDate = dayjs(minimumDate, DATETIME_INPUT_MASK);
	if (clickedDate.startOf('day').isBefore(parsedMinDate)) {
		return parsedMinDate;
	}

	return clickedDate.startOf('day');
};

export const getMaximumDateFromDayClick = (clickedDate: dayjs.Dayjs, maximumDate: string): dayjs.Dayjs => {
	const parsedMaxDate = dayjs(maximumDate, DATETIME_INPUT_MASK);
	if (clickedDate.endOf('day').isAfter(parsedMaxDate)) {
		return parsedMaxDate;
	}

	return clickedDate.endOf('day');
};

export const getFromDateInputState = (error: EAbsoluteTimeError | undefined, { minDate }: IAbsoluteTimeLimits): DateInputState | undefined => {
	if (!error) {
		return;
	}

	if (error === EAbsoluteTimeError.StartDateBeforeMinimumDate && minDate) {
		const min = dayjs(minDate).format(DATETIME_INPUT_MASK);
		return {
			state: EValidationState.Invalid,
			helpText: `The 'FROM' date must be after ${min}`
		};
	}

	return;
};

export const getToDateTimeInputState = (error: EAbsoluteTimeError | undefined, { maxDate }: IAbsoluteTimeLimits): DateInputState | undefined => {
	if (!error) {
		return;
	}

	if (error === EAbsoluteTimeError.EndDateAfterMaximumDate && maxDate) {
		const max = dayjs(maxDate).format(DATETIME_INPUT_MASK);
		return {
			state: EValidationState.Invalid,
			helpText: `The 'TO' date must be before ${max}`
		};
	}

	if (error === EAbsoluteTimeError.EndDateBeforeStartDate) {
		return {
			state: EValidationState.Invalid,
			helpText: `The 'TO' date must be after 'FROM' date`
		};
	}

	return;
};

export const getSingleDateTimeInputState = (error: EAbsoluteTimeError | undefined, { minDate, maxDate }: IAbsoluteTimeLimits): DateInputState | undefined => {
	if (!error) {
		return;
	}

	if (error === EAbsoluteTimeError.StartDateBeforeMinimumDate && minDate) {
		const min = dayjs(minDate).format(DATETIME_INPUT_MASK);
		return {
			state: EValidationState.Invalid,
			helpText: `The date must be after ${min}`
		};
	}

	if (error === EAbsoluteTimeError.EndDateAfterMaximumDate && maxDate) {
		const max = dayjs(maxDate).format(DATETIME_INPUT_MASK);
		return {
			state: EValidationState.Invalid,
			helpText: `The date must be before ${max}`
		};
	}

	return;
};
