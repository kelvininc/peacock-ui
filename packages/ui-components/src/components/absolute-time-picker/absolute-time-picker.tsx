import { Component, Event, EventEmitter, Host, Prop, State, Watch, h } from '@stencil/core';
import { EComponentSize, EIconName, EInputSource } from '../../types';
import dayjs from 'dayjs';
import {
	CALENDAR_DATE_TIME_MASK,
	CALENDAR_INPUT_MAX_DATE,
	CALENDAR_INPUT_MIN_DATE,
	CALENDAR_MASK,
	DATETIME_INPUT_MASK,
	DATE_INPUT_MASK,
	DATE_INPUT_PLACEHOLDER,
	DEFAULT_HEADER_TITLE
} from './absolute-time-picker.config';
import { fromDateInput, fromISO, isDateBefore, isDateSame, newDate } from '../../utils/date.helper';
import { isEmpty } from 'lodash';
import { ERelativeTimeInputMode, IAbsoluteTimePicker, IAbsoluteTimePickerEvents, IRelativeTimeInput, IAbsoluteSelectedRangeDates } from './absolute-time-picker.types';
import { buildSelectedDatesEventPayload, isEndDateAtStartOfDay, isInputDateInLimit } from './absolute-time-picker.helper';
import { IClickDateEvent } from '../time-picker-calendar/time-picker-calendar.types';

@Component({
	tag: 'kv-absolute-time-picker',
	styleUrl: 'absolute-time-picker.scss',
	shadow: false
})
export class KvAbsoluteTimePicker implements IAbsoluteTimePicker, IAbsoluteTimePickerEvents {
	/** @inheritdoc */
	@Prop({ reflect: false }) headerTitle?: string = DEFAULT_HEADER_TITLE;
	/** @inheritdoc */
	@Prop({ reflect: false }) displayBackButton?: boolean = false;
	/** @inheritdoc */
	@Prop({ reflect: false }) selectedRangeDates?: string[] = [];
	/** @inheritdoc */
	@Prop({ reflect: false }) initialDate?: string;
	/** @inheritdoc */
	@Prop({ reflect: false }) disabledDates?: string[] = [];
	/** @inheritdoc */
	@Prop({ reflect: false }) relativeTimeConfig?: IRelativeTimeInput;
	/** @inheritdoc */
	@Prop({ reflect: false }) calendarInputMinDate?: string = CALENDAR_INPUT_MIN_DATE;
	/** @inheritdoc */
	@Prop({ reflect: false }) calendarInputMaxDate?: string = CALENDAR_INPUT_MAX_DATE;

	/** @inheritdoc */
	@Event() selectRangeDatesChange: EventEmitter<IAbsoluteSelectedRangeDates>;
	/** @inheritdoc */
	@Event() backButtonClicked: EventEmitter<MouseEvent>;
	/** @inheritdoc */
	@Event() relativeTimeConfigReset: EventEmitter<MouseEvent>;
	/** @inheritdoc */
	@Event() relativeTimeConfigChange: EventEmitter<IAbsoluteSelectedRangeDates>;

	/** From and to input values inserted by the user on the input */
	@State() fromInputValue: string = '';
	@State() toInputValue: string = '';
	/** Shared hovered date between calendars */
	@State() hoveredDate: string = '';
	/**  First calendar displayed month (Second calendar is displayedMonth + 1) */
	@State() displayedMonth: dayjs.Dayjs = fromDateInput(this.initialDate) ?? fromDateInput(new Date());
	/** Used to force focus the from and to input */
	@State() fromInputFocused: boolean = true;
	@State() toInputFocused: boolean = false;
	/**
	 * Used to define when the input dates came from the relative time picker where the input
	 * can be an absolute date or a text (ex: last 24 hours)
	 */
	@State() inputMode: ERelativeTimeInputMode = ERelativeTimeInputMode.Date;
	/** Calendars max and minimum dates */
	@State() maxDate: string = '';
	@State() minDate: string = '';

	@Watch('selectedRangeDates')
	handleSelectedRangeDatesChange(value: string[] = []) {
		if (value.length === 0) {
			this.resetInputValues();
			this.resetDateLimits();
			return;
		}

		if (isEmpty(this.fromInputValue) && isEmpty(this.toInputValue) && isEmpty(this.relativeTimeConfig)) {
			const [from, to] = value;

			const parsedFromDate = dayjs(from, CALENDAR_DATE_TIME_MASK).format(DATETIME_INPUT_MASK);
			const parsedToDate = dayjs(to, CALENDAR_DATE_TIME_MASK).format(DATETIME_INPUT_MASK);

			this.setInputValues(parsedFromDate, parsedToDate);
		}
	}

	@Watch('relativeTimeConfig')
	handleRelativeTimeConfigInput(newValue: IRelativeTimeInput) {
		if (isEmpty(newValue)) {
			this.inputMode = ERelativeTimeInputMode.Date;
			if (this.selectedRangeDates.length === 0) {
				this.setInputValues('', '');
				this.displayedMonth = fromDateInput(new Date());
			}
			return;
		}

		this.inputMode = newValue.mode;
		this.setInputValues(newValue.from, newValue.to);
		const [from] = this.selectedRangeDates;
		const date = newDate(from);
		if (date.isValid()) {
			this.displayedMonth = date;
		}
	}

	private setDateLimits = (min: string, max: string) => {
		this.minDate = min;
		this.maxDate = max;
	};

	private setInputValues = (from: string, to: string) => {
		this.fromInputValue = from;
		this.toInputValue = to;
	};

	private resetInputValues = () => {
		this.fromInputValue = '';
		this.toInputValue = '';
	};

	private resetDateLimits = () => {
		this.minDate = '';
		this.maxDate = '';
	};

	private emitSelectRangeDatesChangeEvent = (dateA?: dayjs.Dayjs, dateB?: dayjs.Dayjs): void => {
		const range = buildSelectedDatesEventPayload(dateA, dateB);
		this.selectRangeDatesChange.emit({
			range
		});
	};

	private handleBackClick = (event: MouseEvent) => {
		this.backButtonClicked.emit(event);
	};

	private onClickDate = ({ detail }: CustomEvent<IClickDateEvent>): void => {
		const date = detail.date;
		const clickedDate = fromISO(date).startOf('day');
		const inputDate = newDate(date);
		const [selectedStartDate, selectedEndDate] = this.selectedRangeDates;

		if (!selectedStartDate) {
			this.displayedMonth = clickedDate;
			this.setInputValues(inputDate.startOf('day').format(DATETIME_INPUT_MASK), '');
			this.setDateLimits(clickedDate.format(CALENDAR_MASK), '');
			this.emitSelectRangeDatesChangeEvent(clickedDate.startOf('day'));
			return;
		}

		if (isDateSame(clickedDate, selectedStartDate)) {
			if (!selectedEndDate) {
				this.setInputValues(inputDate.startOf('day').format(DATETIME_INPUT_MASK), inputDate.endOf('day').format(DATETIME_INPUT_MASK));
				this.resetDateLimits();
				this.emitSelectRangeDatesChangeEvent(clickedDate.startOf('day'), clickedDate.endOf('day'));

				return;
			}

			this.resetInputValues();
			this.resetDateLimits();
			this.emitSelectRangeDatesChangeEvent();
			return;
		}

		if (selectedEndDate !== undefined) {
			if (!isInputDateInLimit(inputDate)) {
				this.displayedMonth = inputDate;
			}

			this.setInputValues(inputDate.startOf('day').format(DATETIME_INPUT_MASK), '');
			this.setDateLimits(inputDate.format(CALENDAR_MASK), '');
			this.emitSelectRangeDatesChangeEvent(clickedDate.startOf('day'));
			return;
		}

		if (isDateBefore(clickedDate, selectedStartDate)) {
			this.displayedMonth = clickedDate;
			this.setInputValues(inputDate.startOf('day').format(DATETIME_INPUT_MASK), '');
			this.setDateLimits(clickedDate.format(CALENDAR_MASK), '');
			this.emitSelectRangeDatesChangeEvent(clickedDate.startOf('day'));
			return;
		}

		this.setInputValues(newDate(selectedStartDate).format(DATETIME_INPUT_MASK), inputDate.endOf('day').format(DATETIME_INPUT_MASK));
		this.resetDateLimits();
		this.emitSelectRangeDatesChangeEvent(fromISO(selectedStartDate).startOf('day'), clickedDate.endOf('day'));
		return;
	};

	private handleClickBackMonth = () => {
		this.displayedMonth = this.displayedMonth.subtract(1, 'month');
	};

	private handleClickForwardMonth = () => {
		this.displayedMonth = this.displayedMonth.add(1, 'month');
	};

	private handleHoveredDateChange = (event: CustomEvent<string>) => {
		this.hoveredDate = event.detail;
	};

	private getFirstCalendarInitialDate = (): string => {
		const initialDate = this.displayedMonth;
		return initialDate.format(CALENDAR_MASK);
	};

	private getSecondCalendarInitialDate = (): string => {
		const initialDate = this.displayedMonth;
		return initialDate.add(1, 'month').format(CALENDAR_MASK);
	};

	private handleInputReset = () => {
		this.resetInputValues();
		this.inputMode = ERelativeTimeInputMode.Date;
		this.relativeTimeConfigReset.emit();
	};

	private handleOnFocusFromInput = () => {
		this.toInputFocused = false;
		this.fromInputFocused = true;
		const parsedDate = dayjs(this.fromInputValue, DATETIME_INPUT_MASK);

		if (this.relativeTimeConfig) {
			if (this.relativeTimeConfig.mode === ERelativeTimeInputMode.Text) {
				this.handleInputReset();
			} else {
				if (!isEmpty(this.fromInputValue)) {
					this.displayedMonth = parsedDate;
				}
				this.emitSelectRangeDatesChangeEvent(dayjs(this.fromInputValue, DATETIME_INPUT_MASK), dayjs(this.toInputValue, DATETIME_INPUT_MASK));
			}
		} else {
			if (!isEmpty(this.fromInputValue)) {
				this.displayedMonth = parsedDate;
			}
		}
	};

	private handleOnFocusToInput = () => {
		this.toInputFocused = true;
		this.fromInputFocused = false;
		const parsedDate = dayjs(this.toInputValue, DATETIME_INPUT_MASK).subtract(1, 'month');

		if (this.relativeTimeConfig) {
			if (this.relativeTimeConfig.mode === ERelativeTimeInputMode.Text) {
				this.handleInputReset();
			} else {
				if (!isEmpty(this.toInputValue)) {
					this.displayedMonth = parsedDate;
				}
				this.emitSelectRangeDatesChangeEvent(dayjs(this.fromInputValue, DATETIME_INPUT_MASK), dayjs(this.toInputValue, DATETIME_INPUT_MASK));
			}
		} else {
			if (!isEmpty(this.toInputValue)) {
				this.displayedMonth = parsedDate;
			}
		}
	};

	public handleDateChange = (event: CustomEvent<string>, inputSource: EInputSource) => {
		const date = event.detail;
		const parsedDate = dayjs(date, DATE_INPUT_MASK);
		const parsedDateTime = dayjs(date, DATETIME_INPUT_MASK);

		if (parsedDate.isValid() && parsedDateTime.isValid()) {
			if (inputSource === EInputSource.From) {
				this.fromInputValue = parsedDateTime.format(DATETIME_INPUT_MASK);
				this.displayedMonth = parsedDate;
				if (!isEmpty(this.toInputValue)) {
					this.emitSelectRangeDatesChangeEvent(parsedDateTime, dayjs(this.toInputValue, DATETIME_INPUT_MASK));
					this.resetDateLimits();
				} else {
					this.emitSelectRangeDatesChangeEvent(parsedDateTime);
					this.toInputFocused = true;
					this.minDate = parsedDate.format(CALENDAR_MASK);
				}
			} else {
				this.toInputValue = parsedDateTime.format(DATETIME_INPUT_MASK);
				this.displayedMonth = parsedDate.subtract(1, 'month');
				if (!isEmpty(this.fromInputValue)) {
					this.emitSelectRangeDatesChangeEvent(dayjs(this.fromInputValue, DATETIME_INPUT_MASK), parsedDateTime);
					this.resetDateLimits();
				} else {
					this.maxDate = parsedDate.format(CALENDAR_MASK);
					this.emitSelectRangeDatesChangeEvent(parsedDateTime);
					this.fromInputFocused = true;
				}
			}
		} else {
			if (isEmpty(this.fromInputValue)) {
				this.minDate = '';
			}

			if (isEmpty(this.toInputValue)) {
				this.maxDate = '';
			}
		}
	};

	public handleEndDateLostFocus = (event: CustomEvent<string>) => {
		const date = event.detail;
		const parsedDate = dayjs(date, DATE_INPUT_MASK);
		const parsedDateTime = dayjs(date, DATETIME_INPUT_MASK);

		if (parsedDate.isValid() && isEndDateAtStartOfDay(parsedDateTime)) {
			const parsedDateFormated = parsedDate.endOf('day').format(DATETIME_INPUT_MASK);
			this.toInputValue = parsedDateFormated;
			this.displayedMonth = parsedDate.subtract(1, 'month');
			if (!isEmpty(this.fromInputValue)) {
				this.emitSelectRangeDatesChangeEvent(dayjs(this.fromInputValue, DATETIME_INPUT_MASK), parsedDate.endOf('day'));
			} else {
				this.emitSelectRangeDatesChangeEvent(parsedDate.endOf('day'));
				this.fromInputFocused = true;
			}
		}
	};

	/** Components config methods */
	private useInputMask = (): boolean => {
		return this.inputMode === ERelativeTimeInputMode.Date;
	};

	private getInputDateLimits = (source: EInputSource) => {
		if (source === EInputSource.From) {
			const parsedToDateTime = dayjs(this.toInputValue, DATETIME_INPUT_MASK);

			if (this.toInputValue && parsedToDateTime.isValid()) {
				return {
					min: this.calendarInputMinDate,
					max: this.toInputValue
				};
			}
		} else {
			const parsedFromDateTime = dayjs(this.fromInputValue, DATETIME_INPUT_MASK);

			if (this.fromInputValue && parsedFromDateTime.isValid()) {
				return {
					min: this.fromInputValue,
					max: this.calendarInputMaxDate
				};
			}
		}

		return {
			min: this.calendarInputMinDate,
			max: this.calendarInputMaxDate
		};
	};

	render() {
		return (
			<Host>
				<div class="absolute-time-picker-container">
					{this.displayBackButton && (
						<div class="navigate-back" onClick={this.handleBackClick}>
							<kv-icon name={EIconName.SlimRight} customClass="rotate-180" />
							<div class="back-text">Back</div>
						</div>
					)}
					<div class="header">
						<div class="title">{this.headerTitle}</div>
					</div>
					<div class="absolute-input">
						<kv-date-time-input
							useInputMask={this.useInputMask()}
							label="From"
							value={this.fromInputValue}
							size={EComponentSize.Small}
							placeholder={DATE_INPUT_PLACEHOLDER}
							highlighted={isEmpty(this.fromInputValue) && !this.toInputFocused}
							onTextChange={ev => this.handleDateChange(ev, EInputSource.From)}
							onInputFocus={this.handleOnFocusFromInput}
							{...this.getInputDateLimits(EInputSource.From)}
						/>
						<kv-date-time-input
							useInputMask={this.useInputMask()}
							label="To"
							value={this.toInputValue}
							size={EComponentSize.Small}
							placeholder={DATE_INPUT_PLACEHOLDER}
							highlighted={isEmpty(this.toInputValue) && !isEmpty(this.fromInputValue)}
							onTextChange={ev => this.handleDateChange(ev, EInputSource.To)}
							onDateTimeBlur={this.handleEndDateLostFocus}
							onInputFocus={this.handleOnFocusToInput}
							{...this.getInputDateLimits(EInputSource.To)}
						/>
					</div>
					<div class="calendars">
						<kv-time-picker-calendar
							displayNextMonthArrow={false}
							displayPreviousMonthArrow
							selectedDates={this.selectedRangeDates}
							hoveredDate={this.hoveredDate}
							initialDate={this.getFirstCalendarInitialDate()}
							disabledDates={this.disabledDates}
							onClickDate={this.onClickDate}
							onChangeMonth={this.handleClickBackMonth}
							onHoveredDateChange={this.handleHoveredDateChange}
							maxDate={this.maxDate}
							minDate={this.minDate}
						/>
						<kv-time-picker-calendar
							displayNextMonthArrow
							displayPreviousMonthArrow={false}
							selectedDates={this.selectedRangeDates}
							hoveredDate={this.hoveredDate}
							initialDate={this.getSecondCalendarInitialDate()}
							disabledDates={this.disabledDates}
							onClickDate={this.onClickDate}
							onChangeMonth={this.handleClickForwardMonth}
							onHoveredDateChange={this.handleHoveredDateChange}
							maxDate={this.maxDate}
							minDate={this.minDate}
						/>
					</div>
				</div>
			</Host>
		);
	}
}
