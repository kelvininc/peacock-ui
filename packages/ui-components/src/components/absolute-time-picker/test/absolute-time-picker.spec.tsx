import { SpecPage, newSpecPage } from '@stencil/core/testing';
import { KvAbsoluteTimePicker } from '../absolute-time-picker';
import { h } from '@stencil/core';

describe('Absolute Time Picker (unit tests)', () => {
	let page: SpecPage;

	// This component in testing should always consider an initial date
	// to prevent snapshots changes
	describe('when default props and initial date are used', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvAbsoluteTimePicker],
				template: () => <kv-absolute-time-picker initialDate="2023-03-03" />
			});
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});
});
