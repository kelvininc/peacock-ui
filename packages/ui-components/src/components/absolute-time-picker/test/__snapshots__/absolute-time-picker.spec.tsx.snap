// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Absolute Time Picker (unit tests) when default props and initial date are used should match the snapshot 1`] = `
<kv-absolute-time-picker>
  <div class="absolute-time-picker-container">
    <div class="header">
      <div class="title">
        Custom Interval
      </div>
    </div>
    <div class="absolute-range-input">
      <kv-date-time-input highlighted="" inputname="from-input" label="From" placeholder="dd-mm-yyyy 00:00:00" size="small" useinputmask="" value=""></kv-date-time-input>
      <kv-date-time-input inputname="to-input" label="To" placeholder="dd-mm-yyyy 00:00:00" size="small" useinputmask="" value=""></kv-date-time-input>
    </div>
    <div class="calendars">
      <kv-calendar displaypreviousmontharrow="" hovereddate="" initialdate="2023-03-03" maxdate="3000-12-31" mindate="2018-01-01" mode="range"></kv-calendar>
      <kv-calendar displaynextmontharrow="" hovereddate="" initialdate="2023-04-03" maxdate="3000-12-31" mindate="2018-01-01" mode="range"></kv-calendar>
    </div>
  </div>
</kv-absolute-time-picker>
`;
